### ISSUE TITLE
A different take on the idea

### ISSUE THREAD

**Issue by AndyKriger** [geneeblack/graphql_model_mapper#4]
here's the latest form of the GraphQLType module that i originally posted to graphql-ruby

a lot of reorg, cleanup and configurability - there's probably things in there that are oriented towards the project that i'm working on (ex: MetadataInterface) that can be removed or kept or made configurable

[graphql_type.txt](https://github.com/geneeblack/graphql_model_mapper/files/1471157/graphql_type.txt)

i'm not a fan of the explicit create/update/delete macros because i prefer a configurable switch that provides default behavior while still allowing overrides and customizable behavior - keeps everything in one macro with switches to drive behavior

curious what you think and if this is a direction you're interested in going

**Comment by geneeblack** [comment]
I thought originally that the separation of the macros allowed for easier delineation of the functionality and served to provide a straightforward relationship with the underlying model, although I can see where it might be considered too busy on the model and is somewhat self limiting by being tied so closely. Whether it is expressed as individual macros or a single macro with additional parameters is something that would seem to be an easy translation early in the project. 

Right now my thinking is somewhat restricted by my own use cases with the project that I am working. The simplified projection of model functionality serves to keep everything within a defined scope and an abstraction that I'm comfortable with. Everything is easily mappable back to its base model. I had considered supporting aliases of models and attribute names, but at this point could think of no other reason to do it rather than flexibility since there would be no case in my current backing project where it would be required. That's not to say that it would not be useful, just that I've not yet imagined a case where it would be a priority. 

I'm open to bending this into a more flexible implementation, but would like to keep out of the box functionality simplified while providing the capability of extension for those who would like more full featured flexibility.

Let me fully digest what you've put together. After a quick scan it looks like it may allow more flexibility and foster other abstractions of the model.

**Comment by AndyKriger** [comment]
sounds good - i agree with the idea of simple out-of-the-box functionality

i can see how there could be a core implementation along the lines of what i've done that supports the out-of-the-box simplifications that you're aiming for - it'd be cool to be able to use this gem in the project i'm working on but i can only do that if i don't have to change up too much on our end

some things that are in there...
aliases 
type overriding
result formatting
AR enumerations (which get translated into GQL enums)
a resolver wrapper which could support hooks and handles things like translating inputs into something less fragile, translating outputs & errors consistently - in our project, authorization will happen in there too

i don't have specs for it because our testing happens through the specs that use our graphql schema
but it'd be easy enough to write specs that work with the utility methods and stub AR references (like associations or enumerations)

**Comment by geneeblack** [comment]
- aliases: would graphql language aliases get you where you need to be? Or do you see the need to have multiple query/mutation fields addressing the same model, thus requiring aliasing?
- type overriding: which types were you wanting to override? I can see where this would be useful, just not sure where you were wanting to do it.
- result formatting: in the query itself or in the response resolver?
- AR enumerations: support at what level? 
- resolver wrapper: since you can set your own resolver, would that not get you where you want to be? Or were you wanting something deeper or not just on the response?

I think plugin/overriding could be done, just not sure at what level you would want to do it.
As far as authorization/access is concerned what level of granularity does your app require and on which objects is it applied?

Sorry if you addressed some of this in the file you sent previously, I've not had the opportunity to study that yet.

**Comment by geneeblack** [comment]
note to myself, AR Enums are supported in AR >=4.1

**Comment by geneeblack** [comment]
After briefly looking at the file that you sent, it looks like you face similar issues with your metadata field as I face with my Geometry field. I assume you would need to be able to supplement the column mapping from the columns_hash to do Custom type assignment based on that field, not sure if the metadata field's type is consistent or if you need to be able to switch interfaces based on the content of the metadata. Your example implies the latter, that you would need a lazy resolver for that type acted upon query retrieval. 

If that is the case I'll need to think about how that could be accomplished. The type itself with its custom resolver should not be an issue, that could be declared in a custom initializer. Hooking it into the generation step when the columns_hash is mapped is something I'll need to think about. That would require some type of external column->type mapping and I'm not entirely sure of the best way that could be done without some complexity.

**Comment by geneeblack** [comment]
Even though AR enums are not supported in AR 3.2, I don't think it would be a big issue to support them in the OutputTypes when running under Rails 4.1+, possibly even having them in InputTypes would not be an issue. the primary issue crops up in the select query interface since there is no way to use them in a select as a string, since they are resolved in AR not in the DB. That would mean treating them similar to the way scopes are currently treated, but with the additional twist of including comparison values. Not a straight shot, but I think it would be doable.

**Comment by geneeblack** [comment]
@AndyKriger  I'm going to break the Overriding Input/Output attribute GraphQL types part of this discussion out into a separate feature request. #14

 I think this is doable and rather than employ it through configuration, it might be easier to do it through convention. i.e. when scanning the columns_hash if a type is already defined for the column (in an init, i.e. GraphqlModelMapper::[model_name][field_name][input/output?]Type) then the type for that GraphQL field will be assigned to that custom type.

This would keep the options hash simpler, not tie us too deeply to a Rails version and allow custom extension without cluttering base functionality

**Comment by geneeblack** [comment]
@AndyKriger  I'm breaking out the alias request to a separate discussion #15

**Comment by AndyKriger** [comment]
to answer your question about the interface, our metadata type is consistent - it's always a jsonb type so that made it easy to implement

one approach could be custom hooks in the covert_type method if something special was needed to handle type conversion

```
# initialize the Metadata types first
Types::MetadataInterface = GraphQL::InterfaceType.define do
  name 'MetadataInterface'
  description 'Interface definition for any type that can have metadata'
  field :metadata, types[Types::Metadata], 'A list of arbitrary key-value pairs' do
    resolve ->(obj, _, _) { obj.metadata }
  end
end unless defined? Types::MetadataInterface # test for definition to prevent redefinition warning

# this defines a Metadata object
Types::Metadata = GraphQL::ObjectType.define do
  name 'Metadata'
  description 'Generic structure for storing metadata as a map of key-value pairs'
  field :key, types.String, 'The metadata map key'
  field :value, types.String, 'The metadata map value'
end unless defined? Types::Metadata # test for definition to prevent redefinition warning

module Metadata
  # @param metadata [String,Array,Hash] take metadata as a JSON string of key/values, a Hash of key/values, or an Array<Hash<Symbol{key},String{value}>>
  def metadata=(metadata)
    return unless metadata

    self[:metadata] = case metadata
                        when String
                          JSON.parse(metadata) if metadata
                        when Array
                          # turn an Array<Hash> where each Hash is {key:Object, value:Object} into a Hash of key/value
                          # this is the form used to represent metadata in GraphQL
                          Hash[*metadata.map(&:values).flatten]
                        when Hash
                          metadata
                        else
                          raise ArgumentError.new("Invalid type for #{metadata}")
                      end
  end

  # @return metadata as an array of OpenStructs of the form {key:Object, value:Object}
  def metadata
    self[:metadata].map do |key, val|
      OpenStruct.new(key: key, value: val)
    end if self[:metadata]
  end
end
```

**Comment by AndyKriger** [comment]
there was a place where i needed to specify the type was an integer or float for some reason

the result formatting happens in the response resolver wrapper

AR enumerations are handy and translate nicely to GQL enums with the added bonus of value checking by GQL before you even get to the AR model layer

resolver wrapper - we have lots of model classes implementing their own resolvers - i didn't want every resolver to have to concern itself with returning the correct data structure or with generalized error handling - thus the wrapper

**Comment by geneeblack** [comment]
OK, if you look at #14 and PR #16 (branch feature14) I've got a solution for the custom types as well a default handler for enum types aligned to Rails >= 4.1.

So if I understand correctly, you would like to be able to set a global default resolve wrapper, rather than overriding on each type? Somewhat of a replacement or a supplement to the existing global default handlers? I assume it would still be aligned to the top level query/mutation(create/delete/update) types?

**Comment by AndyKriger** [comment]
yes - in the code i've written, there's a global wrapper that calls out to the resolver assigned to a query/mutation

**Comment by geneeblack** [comment]
@AndyKriger breaking out the global resolve wrapper request to #18
